# Workbook

Use this notebook to complete the exercises throughout the workshop.

#### Table of Contents
- [Section 1 &ndash; Getting Started with Pandas](#Section-1)
- [Section 2 &ndash; Data Wrangling](#Section-2)
- [Section 3 &ndash; Data Visualization](#Section-3)

--- 

### Section 1

#### Exercise 1.1
##### Create a DataFrame by reading in the `2019_Yellow_Taxi_Trip_Data.csv` file. Examine the first 5 rows.

In [1]:
import pandas as pd
taxi_trips = pd.read_csv('../data/2019_Yellow_Taxi_Trip_Data.csv')
taxi_trips.head()

vendorid     tpep_pickup_datetime    tpep_dropoff_datetime  \
0         2  2019-10-23T16:39:42.000  2019-10-23T17:14:10.000   
1         1  2019-10-23T16:32:08.000  2019-10-23T16:45:26.000   
2         2  2019-10-23T16:08:44.000  2019-10-23T16:21:11.000   
3         2  2019-10-23T16:22:44.000  2019-10-23T16:43:26.000   
4         2  2019-10-23T16:45:11.000  2019-10-23T16:58:49.000   

   passenger_count  trip_distance  ratecodeid store_and_fwd_flag  \
0                1           7.93           1                  N   
1                1           2.00           1                  N   
2                1           1.36           1                  N   
3                1           1.00           1                  N   
4                1           1.96           1                  N   

   pulocationid  dolocationid  payment_type  fare_amount  extra  mta_tax  \
0           138           170             1         29.5    1.0      0.5   
1            11            26             1         10.5    1.0      0.5   
2           163           162             1          9.5    1.0      0.5   
3           170           163             1         13.0    1.0      0.5   
4           163           236             1         10.5    1.0      0.5   

   tip_amount  tolls_amount  improvement_surcharge  total_amount  \
0        7.98          6.12                    0.3         47.90   
1        0.00          0.00                    0.3         12.30   
2        2.00          0.00                    0.3         15.80   
3        4.32          0.00                    0.3         21.62   
4        0.50          0.00                    0.3         15.30   

   congestion_surcharge  
0                   2.5  
1                   0.0  
2                   2.5  
3                   2.5  
4                   2.5

#### Exercise 1.2
##### Find the dimensions (number of rows and number of columns) in the data.

In [2]:
print("Data dimensions:", taxi_trips.shape[0], "rows,", taxi_trips.shape[1], "columns.")
print("\n")
print("Row data: ", taxi_trips.index)
print("\n")
print("Column data: ", taxi_trips.columns)


Data dimensions: 10000 rows, 18 columns.


Row data:  RangeIndex(start=0, stop=10000, step=1)


Column data:  Index(['vendorid', 'tpep_pickup_datetime', 'tpep_dropoff_datetime',
       'passenger_count', 'trip_distance', 'ratecodeid', 'store_and_fwd_flag',
       'pulocationid', 'dolocationid', 'payment_type', 'fare_amount', 'extra',
       'mta_tax', 'tip_amount', 'tolls_amount', 'improvement_surcharge',
       'total_amount', 'congestion_surcharge'],
      dtype='object')


#### Exercise 1.3
##### Using the data in the `2019_Yellow_Taxi_Trip_Data.csv` file, calculate summary statistics for the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` columns.

In [3]:
taxi_trips.agg(
    {
    "fare_amount": ["min", "max", "median", "mean"],
    "tip_amount": ["min", "max", "median", "mean"],
    "tolls_amount": ["min", "max", "median", "mean"],
    "total_amount": ["min", "max", "median", "mean"] 
    })

fare_amount  tip_amount  tolls_amount  total_amount
min      -52.000000    0.000000     -6.120000    -65.920000
max      176.000000   43.000000    612.000000    671.800000
median    10.000000    2.000000      0.000000     16.300000
mean      15.106313    2.634494      0.623447     22.564659

#### Exercise 1.4
##### Isolate the `fare_amount`, `tip_amount`, `tolls_amount`, and `total_amount` for the longest trip by distance (`trip_distance`).

In [4]:
# max_trip_distance_data = taxi_trips.loc[taxi_trips['trip_distance'].idxmax()]
max_dist = taxi_trips['trip_distance'].idxmax()

print("Max Distance Data:")
print("Fare:", taxi_trips['fare_amount'][max_dist])
print("Tip:", taxi_trips['tip_amount'][max_dist])
print("Tolls:", taxi_trips['tip_amount'][max_dist])
print("Total:", taxi_trips['total_amount'][max_dist])


Max Distance Data:
Fare: 176.0
Tip: 18.29
Tolls: 18.29
Total: 201.21


---

### Section 2

#### Exercise 2.1
##### Read in the meteorite data from the `Meteorite_Landings.csv` file, rename the `mass (g)` column to `mass`, and drop all the latitude and longitude columns. Sort the result by mass in descending order.

In [127]:
meteorites = pd.read_csv('../data/Meteorite_Landings.csv')

meteorites = meteorites.sort_values(by = 'mass (g)', ascending = False)
meteorites = meteorites.drop(['reclat', 'reclong'], axis=1)
meteorites = meteorites.rename(columns = {'mass (g)': 'mass'})

#meteorites


#### Exercise 2.2
##### Using the meteorite data from the `Meteorite_Landings.csv` file, update the `year` column to only contain the year, convert it to a numeric data type, and create a new column indicating whether the meteorite was observed falling before 1970. Set the index to the `id` column and extract all the rows with IDs between 10,036 and 10,040 (inclusive) with `loc[]`.

###### **Hint 1**: Use `year.str.slice()` to grab a substring.

###### **Hint 2**: Make sure to sort the index before using `loc[]` to select the range.

###### **Bonus**: There's a data entry error in the `year` column. Can you find it? (Don't spend too much time on this.)

In [128]:
meteorites = meteorites.dropna(subset='year')
meteorites['year'] = meteorites['year'].str.slice(start=6, stop=10)
meteorites['year'] = pd.to_numeric(meteorites['year'], downcast = 'integer')
# Want to figure out how to reorder the columns so that the new one is after 'year'
meteorites['Before 1970?'] = np.where(meteorites['year'] < 1970, 'Yes', 'No')
meteorites

name     id nametype      recclass        mass   fall  year  \
16392             Hoba  11890    Valid     Iron, IVB  60000000.0  Found  1920   
5373         Cape York   5262    Valid   Iron, IIIAB  58200000.0  Found  1818   
5365   Campo del Cielo   5247    Valid  Iron, IAB-MG  50000000.0  Found  1575   
5370     Canyon Diablo   5257    Valid  Iron, IAB-MG  30000000.0  Found  1891   
3455           Armanty   2335    Valid    Iron, IIIE  28000000.0  Found  1898   
...                ...    ...      ...           ...         ...    ...   ...   
38282   Wei-hui-fu (a)  24231    Valid          Iron         NaN  Found  1931   
38283   Wei-hui-fu (b)  24232    Valid          Iron         NaN  Found  1931   
38285          Weiyuan  24233    Valid  Mesosiderite         NaN  Found  1978   
41472    Yamato 792768  28117    Valid           CM2         NaN  Found  1979   
45698    Zapata County  30393    Valid          Iron         NaN  Found  1930   

                  GeoLocation Before 1970?  
16392   (-19.58333, 17.91667)          Yes  
5373    (76.13333, -64.93333)          Yes  
5365   (-27.46667, -60.58333)          Yes  
5370      (35.05, -111.03333)          Yes  
3455             (47.0, 88.0)          Yes  
...                       ...          ...  
38282                     NaN          Yes  
38283                     NaN          Yes  
38285   (35.26667, 104.31667)           No  
41472       (-71.5, 35.66667)           No  
45698           (27.0, -99.0)          Yes  

[45425 rows x 9 columns]

#### Exercise 2.3
##### Using the meteorite data from the `Meteorite_Landings.csv` file, create a pivot table that shows both the number of meteorites and the 95th percentile of meteorite mass for those that were found versus observed falling per year from 2005 through 2009 (inclusive). Hint: Be sure to convert the `year` column to a number as we did in the previous exercise.

In [159]:
# Calculate number of meteorites that fell in a given year
met_mass = meteorites.dropna(subset='mass')

met_count = met_mass.value_counts('year').to_frame('#')
met_count = met_count.sort_values(by = 'year', ascending = True)
met_observed = met_mass.value_counts(['year', 'fall'])
met_observed

#meteorites_95 = meteorites.pivot_table(
#    index = ['year', 'mnum'] , columns = 'mass', sort = False
#)

#meteorites_95

#
year      
860      1
1399     1
1490     1
1491     1
1575     1
...    ...
2010  1005
2011   713
2012   234
2013    11
2101     1

[253 rows x 1 columns]

#### Exercise 2.4
##### Using the meteorite data from the `Meteorite_Landings.csv` file, compare summary statistics of the mass column for the meteorites that were found versus observed falling.

#### Exercise 2.5
##### Using the taxi trip data in the `2019_Yellow_Taxi_Trip_Data.csv` file, resample the data to an hourly frequency based on the dropoff time. Calculate the total `trip_distance`, `fare_amount`, `tolls_amount`, and `tip_amount`, then find the 5 hours with the most tips.

--- 

### Section 3

#### Exercise 3.1
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create box plots for traveler throughput for each year in the data. Hint: Pass `kind='box'` into the `plot()` method to generate box plots.

#### Exercise 3.2
##### Using the TSA traveler throughput data in the `tsa_melted_holiday_travel.csv` file, create a heatmap that shows the 2019 TSA median traveler throughput by day of week and month.

#### Exercise 3.3
##### Annotate the medians in the box plot from *[Exercise 3.1](#Exercise-3.1)*. Hint: The `x` coordinates will be 1, 2, and 3 for 2019, 2020, and 2021, respectively. Alternatively, to avoid hardcoding values, you can use the `Axes.get_xticklabels()` method, in which case you should look at the [documentation](https://matplotlib.org/stable/api/text_api.html) for the `Text` class.